src/tree.rs

```diff
@@ -101,3 +101,138 @@ fn hash_layer(l_star: usize, pair: Pair) -> pallas::Base {
         )
         .unwrap()
 }
+
+/// Generators for property testing.
+#[cfg(any(test, feature = "test-dependencies"))]
+pub mod testing {
+    use lazy_static::lazy_static;
+    use std::convert::TryInto;
+
+    use crate::{
+        constants::{MERKLE_CRH_PERSONALIZATION, MERKLE_DEPTH_ORCHARD},
+        note::{commitment::ExtractedNoteCommitment, testing::arb_note, Note},
+        primitives::sinsemilla::{i2lebsp_k, HashDomain, K},
+        value::{testing::arb_positive_note_value, MAX_NOTE_VALUE},
+    };
+    use ff::{PrimeField, PrimeFieldBits};
+    use pasta_curves::pallas;
+
+    use proptest::collection::vec;
+    use proptest::prelude::*;
+
+    use super::{hash_layer, Anchor, MerklePath, Pair};
+
+    lazy_static! {
+        static ref EMPTY_ROOTS: Vec<pallas::Base> = {
+            // MerkleCRH Sinsemilla hash domain.
+            let domain = HashDomain::new(MERKLE_CRH_PERSONALIZATION);
+
+            let mut v = vec![pallas::Base::zero()];
+            for l_star in 0..MERKLE_DEPTH_ORCHARD {
+                let next = domain
+                .hash(
+                    std::iter::empty()
+                        .chain(i2lebsp_k(l_star).iter().copied().take(K))
+                        .chain(
+                            v[l_star]
+                            .to_le_bits()
+                            .iter()
+                            .by_val()
+                            .take(pallas::Base::NUM_BITS as usize)
+                        )
+                        .chain(
+                            v[l_star]
+                            .to_le_bits()
+                            .iter()
+                            .by_val()
+                            .take(pallas::Base::NUM_BITS as usize)
+                        ),
+                )
+                .unwrap();
+                v.push(next);
+            }
+            v
```

Comment:
This can just use `hash_layer`. It could also be implemented with a scan:
```suggestion
            Some(pallas::Base::zero())
                .chain((0..MERKLE_DEPTH_ORCHARD).scan(pallas::Base::zero(), |state, l_star| {
                    *state = hash_layer(l_star, Pair { left: *state, right: *state });
                    Some(*state)
                })
                .collect()
```